File: nim/aoc_solutions/2017/day_03.nim
import std/[strutils, tables, math]
import aoclib/aoc_input

const
  Year* = 2017
  Day*  = 3

type
  Coord = tuple[x, y: int]

const
  turns: array[4, Coord] = [
    (x:  1, y:  0),  # right
    (x:  0, y:  1),  # up
    (x: -1, y:  0),  # left
    (x:  0, y: -1)   # down
  ]

  neighborOffsets: array[8, Coord] = [
    (x: -1, y:  1), (x:  0, y:  1), (x:  1, y:  1),
    (x: -1, y:  0),                  (x:  1, y:  0),
    (x: -1, y: -1), (x:  0, y: -1), (x:  1, y: -1)
  ]

proc getCoordinateOf(target: int): Coord =
  ## Ritorna le coordinate cartesiane (x,y) della cella "target" nella spirale.
  if target == 1:
    return (x: 0, y: 0)

  var
    turnIndex = 0
    distance  = 1
    value     = 1
    steps     = 0
    actual    = (x: 0, y: 0)

  while true:
    let turn = turns[turnIndex]
    let nextPos = (
      x: actual.x + turn.x * distance,
      y: actual.y + turn.y * distance
    )
    actual = nextPos

    value += distance
    inc steps

    # prossimo lato della spirale
    turnIndex = (turnIndex + 1) mod 4
    if steps mod 2 == 0:
      inc distance

    if value >= target:
      let delta = value - target
      result = (
        x: actual.x - turn.x * delta,
        y: actual.y - turn.y * delta
      )
      break

proc tryGetAdj(grid: Table[Coord, int], x, y: int): int =
  var sum = 0
  for off in neighborOffsets:
    let pos = (x: x + off.x, y: y + off.y)
    if grid.hasKey(pos):
      sum += grid[pos]
  result = sum

proc getFirstLarger(target: int): int =
  var
    turnIndex = 0
    distance  = 1
    value     = 1
    steps     = 0
    actual    = (x: 0, y: 0)
    grid      = initTable[Coord, int]()

  grid[actual] = 1

  while true:
    let turn = turns[turnIndex]
    for _ in 0 ..< distance:
      let nextPos = (
        x: actual.x + turn.x,
        y: actual.y + turn.y
      )
      actual = nextPos

      value = tryGetAdj(grid, actual.x, actual.y)
      if value > target:
        result = value
        return

      grid[actual] = value

    inc steps
    turnIndex = (turnIndex + 1) mod 4
    if steps mod 2 == 0:
      inc distance

proc part1*(input: string): string =
  let target = input.strip.parseInt
  let c = getCoordinateOf(target)
  let dist = abs(c.x) + abs(c.y)
  result = $dist

proc part2*(input: string): string =
  let target = input.strip.parseInt
  let v = getFirstLarger(target)
  result = $v

when isMainModule:
  let input: string = readInput(Year, Day)
  echo "Part 1: ", part1(input)
  echo "Part 2: ", part2(input)

File: python/aoc_solutions/2017/day_03.py
from __future__ import annotations

from pathlib import Path
import sys
from dataclasses import dataclass


PYTHON_DIR = Path(__file__).resolve().parents[2]
if str(PYTHON_DIR) not in sys.path:
    sys.path.append(str(PYTHON_DIR))

from get_input import GetInput


GI = GetInput()



def get_coordinate_of(target: int) -> tuple[int, int]:
    turns = [(1,0), (0,1), (-1,0), (0,-1)]
    turn_index = 0
    distance = 1
    value = 1
    actualx, actualy = 0,0
    steps = 0

    while True:
        turn = turns[turn_index]
        nextx, nexty = actualx + turn[0]*distance, actualy + turn[1]*distance
        actualx, actualy = nextx, nexty

        value += distance
        steps += 1

        turn_index += 1
        turn_index = turn_index % 4

        if steps % 2 == 0:
            distance += 1
        
        if value >= target:
            delta = value - target
            result_x, result_y = actualx - turn[0]*delta, actualy - turn[1]*delta
            return result_x, result_y

def try_get_adj(D: dict, x: int, y: int)->int:
    P = [[x-1,y+1] , [x,y+1], [x+1, y+1],
         [x-1, y]  ,          [x+1, y],
         [x-1, y-1], [x, y-1], [x+1, y-1]
         ]
    res = 0
    for p in P:
        try:
            res += D[(p[0],p[1])]
        except:
            pass
    
    return res

def get_first_larger(target: int) -> tuple[int, int, int]:
    turns = [(1,0), (0,1), (-1,0), (0,-1)]
    turn_index = 0
    distance = 1
    value = 1
    actualx, actualy = 0,0
    steps = 0

    D: dict[tuple[int, int], int] = dict()
    D[(0,0)] = 1

    while True:
        turn = turns[turn_index]
        for i in range(distance):
            nextx, nexty = actualx + turn[0], actualy + turn[1]
            actualx, actualy = nextx, nexty

            value = try_get_adj(D, actualx, actualy)
            if value > target:
                return value, actualx, actualy
            
            D[(actualx, actualy)] = value

        steps += 1

        turn_index += 1
        turn_index = turn_index % 4

        if steps % 2 == 0:
            distance += 1


def solve_1(test_string: str | None = None) -> int:
    inputs_1 = GI.input if test_string is None else test_string

    coordinates = get_coordinate_of(int(inputs_1))
    print(coordinates)

    return abs(coordinates[1]) + abs(coordinates[0])


def solve_2(test_string: str | None = None) -> int:
    inputs_1 = GI.input if test_string is None else test_string

    value, x, y = get_first_larger(int(inputs_1))
    return value


if __name__ == "__main__":
    print(f"Part 1: {solve_1()}")
    print(f"Part 2: {solve_2()}")
